File: macos/Tests/AudioTranscriptionSummaryTests/AdditionalPromptPersistenceTests.swift
// AdditionalPromptPersistenceTests.swift
// 追加プロンプトの設定保存・復元テスト
// - 要約実行時に追加プロンプトが設定 JSON に保存されること
// - アプリ起動時（AppViewModel 初期化時）に復元されること

import Testing
import Foundation
@testable import AudioTranscriptionSummary

/// テストデータディレクトリのパス
private let testDataDir: URL = {
    var url = URL(fileURLWithPath: #filePath)
    for _ in 0..<4 { url = url.deletingLastPathComponent() }
    return url.appendingPathComponent("test/data")
}()

@Suite("追加プロンプト永続化テスト")
struct AdditionalPromptPersistenceTests {

    @Test("追加プロンプトが設定 JSON に保存・復元される")
    func testSaveAndRestore() throws {
        let settingsDir = testDataDir.appendingPathComponent("settings_test")
        let settingsFile = settingsDir.appendingPathComponent("settings.json")
        defer {
            try? FileManager.default.removeItem(at: settingsDir)
        }

        let store = AppSettingsStore(directory: settingsDir)

        // 初期状態: 空文字
        let initial = store.load()
        #expect(initial.summaryAdditionalPrompt == "")

        // 保存
        var settings = initial
        settings.summaryAdditionalPrompt = "箇条書きで要約して"
        try store.save(settings)

        // ファイルが生成されたことを確認
        #expect(FileManager.default.fileExists(atPath: settingsFile.path))

        // 復元
        let restored = store.load()
        #expect(restored.summaryAdditionalPrompt == "箇条書きで要約して")
    }

    @Test("空文字の追加プロンプトも正しく保存・復元される")
    func testEmptyPrompt() throws {
        let settingsDir = testDataDir.appendingPathComponent("settings_empty_test")
        defer {
            try? FileManager.default.removeItem(at: settingsDir)
        }

        let store = AppSettingsStore(directory: settingsDir)

        var settings = store.load()
        settings.summaryAdditionalPrompt = "テスト"
        try store.save(settings)

        // 空文字で上書き
        settings.summaryAdditionalPrompt = ""
        try store.save(settings)

        let restored = store.load()
        #expect(restored.summaryAdditionalPrompt == "")
    }

    @Test("他の設定値に影響しない")
    func testNoSideEffects() throws {
        let settingsDir = testDataDir.appendingPathComponent("settings_sideeffect_test")
        defer {
            try? FileManager.default.removeItem(at: settingsDir)
        }

        let store = AppSettingsStore(directory: settingsDir)

        var settings = store.load()
        settings.region = "us-east-1"
        settings.bedrockModelId = "anthropic.claude-sonnet-4-6"
        settings.summaryAdditionalPrompt = "テストプロンプト"
        try store.save(settings)

        let restored = store.load()
        #expect(restored.region == "us-east-1")
        #expect(restored.bedrockModelId == "anthropic.claude-sonnet-4-6")
        #expect(restored.summaryAdditionalPrompt == "テストプロンプト")
    }
}

File: test/windows/AdditionalPromptPersistenceTests.cs
// AdditionalPromptPersistenceTests.cs
// 追加プロンプトの設定保存・復元テスト
// - 要約実行時に追加プロンプトが設定 JSON に保存されること
// - アプリ起動時に復元されること
//
// テスト生成ファイルは test/data/ 以下に出力し、テスト完了後に削除する

using System;
using System.IO;
using System.Text.Json;
using Xunit;
using AudioTranscriptionSummary.Models;

namespace AudioTranscriptionSummary.Tests;

public class AdditionalPromptPersistenceTests : IDisposable
{
    private static readonly string TestDataDir = Path.GetFullPath(
        Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "..", "test", "data"));

    private readonly string _settingsDir;
    private readonly string _settingsFile;

    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    public AdditionalPromptPersistenceTests()
    {
        _settingsDir = Path.Combine(TestDataDir, $"settings_test_{Guid.NewGuid():N}");
        _settingsFile = Path.Combine(_settingsDir, "settings.json");
        Directory.CreateDirectory(_settingsDir);
    }

    public void Dispose()
    {
        if (Directory.Exists(_settingsDir))
            Directory.Delete(_settingsDir, true);
    }

    private AppSettings LoadSettings()
    {
        if (!File.Exists(_settingsFile))
            return new AppSettings();
        var json = File.ReadAllText(_settingsFile);
        return JsonSerializer.Deserialize<AppSettings>(json, JsonOptions) ?? new AppSettings();
    }

    private void SaveSettings(AppSettings settings)
    {
        var json = JsonSerializer.Serialize(settings, JsonOptions);
        File.WriteAllText(_settingsFile, json);
    }

    [Fact]
    public void SaveAndRestore_AdditionalPrompt()
    {
        // 初期状態: 空文字
        var initial = LoadSettings();
        Assert.Equal("", initial.SummaryAdditionalPrompt);

        // 保存
        initial.SummaryAdditionalPrompt = "箇条書きで要約して";
        SaveSettings(initial);

        Assert.True(File.Exists(_settingsFile));

        // 復元
        var restored = LoadSettings();
        Assert.Equal("箇条書きで要約して", restored.SummaryAdditionalPrompt);
    }

    [Fact]
    public void EmptyPrompt_SavedAndRestoredCorrectly()
    {
        var settings = new AppSettings { SummaryAdditionalPrompt = "テスト" };
        SaveSettings(settings);

        // 空文字で上書き
        settings.SummaryAdditionalPrompt = "";
        SaveSettings(settings);

        var restored = LoadSettings();
        Assert.Equal("", restored.SummaryAdditionalPrompt);
    }

    [Fact]
    public void NoSideEffects_OnOtherSettings()
    {
        var settings = new AppSettings
        {
            Region = "us-east-1",
            BedrockModelId = "anthropic.claude-sonnet-4-6",
            SummaryAdditionalPrompt = "テストプロンプト"
        };
        SaveSettings(settings);

        var restored = LoadSettings();
        Assert.Equal("us-east-1", restored.Region);
        Assert.Equal("anthropic.claude-sonnet-4-6", restored.BedrockModelId);
        Assert.Equal("テストプロンプト", restored.SummaryAdditionalPrompt);
    }
}
